### PR TITLE
Enum To String using Magic_Enum Library

### LINKED ISSUE #857: Implement A Method For Defining An Enum Alongside Its String Representation
### Description of the task
Currently, if we want the string representation for an enum, we currently just define a lookup table from the enum name to the string (see below for example). This is problematic for several reasons:
a) it's extra work / overhead to manage when the enum is changed, ensuring the table is updated and tested, etc. etc.
b) It's error-prone, easy to forget to add an entry to the table, or copy-pasta an entry and forget to change the actual string.

We should have some way of defining an enum alongside it's string representation that we can use anywhere in the repository. One possible good option could be [this stackoverflow answer](https://stackoverflow.com/questions/207976/how-to-easily-map-c-enums-to-strings/23404302#23404302) or if we want to use an external library, then the first answer on [this stackoverflow post](https://stackoverflow.com/questions/28828957/enum-to-string-in-modern-c11-c14-c17-and-future-c20).

 It's probably worth having a discussion about where this should go, so talk to @garethellis0 or @MathewMacDougall when you start this.

ex:
`refbox_constants.h`
```
enum class RefboxGameState
{
    // these enum items map to the constants in RefboxCommand.msg
    HALT                 = 0,
    STOP                 = 1,
    NORMAL_START         = 2,
    FORCE_START          = 3,
    PREPARE_KICKOFF_US   = 4,
    PREPARE_KICKOFF_THEM = 5,
    PREPARE_PENALTY_US   = 6,
    PREPARE_PENALTY_THEM = 7,
    DIRECT_FREE_US       = 8,
    DIRECT_FREE_THEM     = 9,
    INDIRECT_FREE_US     = 10,
    INDIRECT_FREE_THEM   = 11,
    TIMEOUT_US           = 12,
    TIMEOUT_THEM         = 13,
    GOAL_US              = 14,
    GOAL_THEM            = 15,
    BALL_PLACEMENT_US    = 16,
    BALL_PLACEMENT_THEM  = 17,
    REFBOX_GAME_STATE_COUNT
};

/**
 * Returns the name of the given refbox gamestate
 *
 * @param state The refbox gamestate to get the name of
 * @return The name of the given refbox gamestate
 */
std::string name(const RefboxGameState& state);
std::ostream& operator<<(std::ostream& os, const RefboxGameState& state);
```
`refbox_constants.cpp`
```
static const std::map<RefboxGameState, std::string> refbox_game_state_names = {
    {RefboxGameState::HALT, "HALT"},
    {RefboxGameState::STOP, "STOP"},
    {RefboxGameState::NORMAL_START, "NORMAL_START"},
    {RefboxGameState::FORCE_START, "FORCE_START"},
    {RefboxGameState::PREPARE_KICKOFF_US, "PREPARE_KICKOFF_US"},
    {RefboxGameState::PREPARE_KICKOFF_THEM, "PREPARE_KICKOFF_THEM"},
    {RefboxGameState::PREPARE_PENALTY_US, "PREPARE_PENALTY_US"},
    {RefboxGameState::PREPARE_PENALTY_THEM, "PREPARE_PENALTY_THEM"},
    {RefboxGameState::DIRECT_FREE_US, "DIRECT_FREE_US"},
    {RefboxGameState::DIRECT_FREE_THEM, "DIRECT_FREE_THEM"},
    {RefboxGameState::INDIRECT_FREE_US, "INDIRECT_FREE_US"},
    {RefboxGameState::INDIRECT_FREE_THEM, "INDIRECT_FREE_THEM"},
    {RefboxGameState::TIMEOUT_US, "TIMEOUT_US"},
    {RefboxGameState::TIMEOUT_THEM, "TIMEOUT_THEM"},
    {RefboxGameState::GOAL_US, "GOAL_US"},
    {RefboxGameState::GOAL_THEM, "GOAL_THEM"},
    {RefboxGameState::BALL_PLACEMENT_US, "BALL_PLACEMENT_US"},
    {RefboxGameState::BALL_PLACEMENT_THEM, "BALL_PLACEMENT_THEM"},
    {RefboxGameState::REFBOX_GAME_STATE_COUNT, "REFBOX_GAME_STATE_COUNT"}};

std::string name(const RefboxGameState& state)
{
    auto find_name_result = refbox_game_state_names.find(state);
    if (find_name_result != refbox_game_state_names.end())
    {
        return find_name_result->second;
    }
    else
    {
        throw std::invalid_argument(
            "Error: Tried to get the name of an invalid refbox gamestate");
    }
}

std::ostream& operator<<(std::ostream& os, const RefboxGameState& state)
{
    if (refbox_game_state_names.find(state) != refbox_game_state_names.end())
    {
        os << refbox_game_state_names.find(state)->second;
    }
    else
    {
        os << "INVALID RefboxGameState: " << static_cast<int>(state);
    }
    return os;
}
```

### Acceptance criteria

- [x] Come up with, and implement, some method that allows us to define an enum alongside its string representation, without needing to do any specific work for each enum entry.
- [ ] Convert all places where we currently use a lookup table to use the new method (hint: try searching for anything that matches `std::map<.*, std::string>`)

### PR DESCRIPTION
**Pull request by kelvinkoon**
### Description
Certain places convert from enums to strings using a lookup table. This causes unnecessary overhead when changes need to be made. The magic_enum library lets us convert from enum to string directly. Changes have been made to:
- refbox_constants
- avoid_area
- motion_constraints

### Testing Done
The refbox_constant_test.cpp passes so far.

### Resolved Issues
Resolves #857 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/world/refbox_constants.cpp [hunk 1]
```diff
@@ -1,51 +1,11 @@
...
-        os << refbox_game_state_names.find(state)->second;
-    }
-    else
-    {
-        os << "INVALID RefboxGameState: " << static_cast<int>(state);
-    }
-    return os;
-}
+    return find_name_result;
+}
```
Comment (garethellis0) [inline src/software/world/refbox_constants.cpp]: nit; missing newline at end of file

File: src/software/world/refbox_constants.h [hunk 2]
```diff
@@ -32,5 +32,4 @@ enum class RefboxGameState
  * @param state The refbox gamestate to get the name of
  * @return The name of the given refbox gamestate
  */
-std::string name(const RefboxGameState& state);
-std::ostream& operator<<(std::ostream& os, const RefboxGameState& state);
\ No newline at end of file
+std::string name(const RefboxGameState& state);
```
Comment (garethellis0) [inline src/software/world/refbox_constants.h]: nit; missing newline at EOF

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Wow, this is beautiful :heart_eyes:

File: src/software/world/refbox_constants.cpp [hunk 3]
```diff
@@ -1,51 +1,11 @@
...
-    else
-    {
-        throw std::invalid_argument(
-            "Error: Tried to get the name of an invalid refbox gamestate");
-    }
-}
+    std::string find_name_result = static_cast<std::string>(magic_enum::enum_name(state));
 
-std::ostream& operator<<(std::ostream& os, const RefboxGameState& state)
-{
```
Comment (jonathanlew) [inline src/software/world/refbox_constants.cpp]: Please keep this operator and have it call name()

**Review by jonathanlew — CHANGES_REQUESTED** [review]
It looks very clean! 
Can you also implement this for motion_constraints.cpp for the "<<" operator

File: src/software/world/refbox_constants.cpp (hunk 3 above)
Comment (kelvinkoon) [reply src/software/world/refbox_constants.cpp]: Added back. Also implemented for motion_constraints.cpp, where it is passing in the motion_constraints_manager_test.cpp.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Just one nit, but this looks _so much cleaner_ :pray:

**Review by jonathanlew — DISMISSED** [review]
LGTM! Thanks for fixing motion constraints

**Comment by kelvinkoon** [comment]
> LGTM! Thanks for fixing motion constraints

Still need to find all the places that do the lookup table and replace it, but I'm glad everyone likes the changes :+1:

**Review by jonathanlew — APPROVED** [review]
Good work! Looks good to me!

**Comment by garethellis0** [comment]
@kelvinkoon - just as a note for the future, please leave a quick reply to each of my comments, but leave them unresolved. This helps when I come back to re-review PR's, especially larger ones.

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
:tada:

**Pull request merged**